Source/JavaScriptCore/runtime/JSGlobalObjectDebuggable.cpp

```diff
@@ -35,49 +35,98 @@
 #include "JSLock.h"
 #include "RemoteInspector.h"
 #include <wtf/TZoneMallocInlines.h>
+#include <wtf/threads/BinarySemaphore.h>
 
 using namespace Inspector;
 
 namespace JSC {
 
 WTF_MAKE_TZONE_ALLOCATED_IMPL(JSGlobalObjectDebuggable);
 
+Ref<JSGlobalObjectDebuggable> JSGlobalObjectDebuggable::create(JSGlobalObject& globalObject)
+{
+    return adoptRef(*new JSGlobalObjectDebuggable(globalObject));
+}
+
 JSGlobalObjectDebuggable::JSGlobalObjectDebuggable(JSGlobalObject& globalObject)
-    : m_globalObject(globalObject)
+    : m_globalObject(&globalObject)
+    , m_globalObjectRunLoop(RunLoop::current())
+{
+}
+
+void JSGlobalObjectDebuggable::callOnGlobalObjectRunLoopAndWait(Function<void()>&& function) const
 {
+    if (&RunLoop::current() == m_globalObjectRunLoop.ptr()) {
+        function();
+        return;
+    }
+    BinarySemaphore semaphore;
+    m_globalObjectRunLoop->dispatch([function = WTFMove(function), &semaphore] {
+        function();
+        semaphore.signal();
+    });
+    semaphore.wait();
 }
 
 String JSGlobalObjectDebuggable::name() const
 {
-    String name = m_globalObject.name();
+    String name;
+    callOnGlobalObjectRunLoopAndWait([this, protectedThis = Ref { *this }, &name] {
+        if (m_globalObject)
+            name = m_globalObject->name().isolatedCopy();
+    });
     return name.isEmpty() ? "JSContext"_s : name;
 }
 
 void JSGlobalObjectDebuggable::connect(FrontendChannel& frontendChannel, bool automaticInspection, bool immediatelyPause)
 {
-    JSLockHolder locker(&m_globalObject.vm());
+    callOnGlobalObjectRunLoopAndWait([this, protectedThis = Ref { *this }, &frontendChannel, automaticInspection, immediatelyPause] {
+        if (!m_globalObject)
+            return;
 
-    m_globalObject.inspectorController().connectFrontend(frontendChannel, automaticInspection, immediatelyPause);
+        JSLockHolder locker(&m_globalObject->vm());
+        m_globalObject->inspectorController().connectFrontend(frontendChannel, automaticInspection, immediatelyPause);
+    });
 }
 
 void JSGlobalObjectDebuggable::disconnect(FrontendChannel& frontendChannel)
 {
-    JSLockHolder locker(&m_globalObject.vm());
+    callOnGlobalObjectRunLoopAndWait([this, protectedThis = Ref { *this }, &frontendChannel] {
+        if (!m_globalObject)
+            return;
 
-    m_globalObject.inspectorController().disconnectFrontend(frontendChannel);
+        JSLockHolder locker(&m_globalObject->vm());
+
+        m_globalObject->inspectorController().disconnectFrontend(frontendChannel);
+    });
 }
 
 void JSGlobalObjectDebuggable::dispatchMessageFromRemote(String&& message)
 {
-    JSLockHolder locker(&m_globalObject.vm());
+    callOnGlobalObjectRunLoopAndWait([this, protectedThis = Ref { *this }, message = WTFMove(message).isolatedCopy()] {
```

Comment:
`callOnGlobalObjectRunLoopAndWait([this, protectedThis = Ref { *this }, message = WTFMove(message).isolatedCopy()]() mutable {`